### PR TITLE
Clean chipnet

### PR DESCRIPTION
**Pull request by rishabh-16**

### CODE REVIEW

File: trailmet/algorithms/prune/chipnet.py [hunk 1]
```diff
@@ -24,34 +27,37 @@ def fo_hook(module, in_tensor, out_tensor):
                 module.output_area = out_tensor.size(2) * out_tensor.size(3)
             conv_module.register_forward_hook(fo_hook)
         self._conv_module = conv_module
-        beta=1.
-        gamma=2.
+        beta = 1.
```
Comment (dkgupta90) [inline trailmet/algorithms/prune/chipnet.py:30]: why do we have hyperparameters defined in here? shouldn't these go as part of the config file? as part of something like chipnet_params?

File: trailmet/algorithms/prune/chipnet.py [hunk 2]
```diff
@@ -24,34 +27,37 @@ def fo_hook(module, in_tensor, out_tensor):
...
-        return self.__continous_heavy_side(zeta_i)
+        return self.__continuous_heavy_side(zeta_i)
 
     def set_beta_gamma(self, beta, gamma):
+        """sets the values of beta and gamma"""
         self.beta.data.copy_(torch.Tensor([beta]))
         self.gamma.data.copy_(torch.Tensor([gamma]))
-      
+
     def __generalized_logistic(self, x):
```
Comment (dkgupta90) [inline trailmet/algorithms/prune/chipnet.py:55]: doesnt a method from standad pytorch exist for this? if not, i think we need to move it to some class called Projections or something similar?this way we can experiment with different projection functions.

File: trailmet/algorithms/prune/chipnet.py [hunk 3]
```diff
@@ -24,34 +27,37 @@ def fo_hook(module, in_tensor, out_tensor):
...
         return 1./(1.+torch.exp(-self.beta*x))
-    
-    def __continous_heavy_side(self, x):
+
+    def __continuous_heavy_side(self, x):
```
Comment (dkgupta90) [inline trailmet/algorithms/prune/chipnet.py]: heavy side is to be spelled as heaviside. it is a popular word.

File: trailmet/algorithms/prune/chipnet.py [hunk 4]
```diff
@@ -72,7 +78,7 @@ def get_volume(self):
         total_volume = self._conv_module.output_area*self.num_gates
         active_volume = self._conv_module.output_area*self.pruned_zeta.sum().item()
         return active_volume, total_volume
-    
+
     def get_flops(self):
```
Comment (dkgupta90) [inline trailmet/algorithms/prune/chipnet.py:82]: this seems to me a generic function. Can we somehow take it out of here into a class in utils for flop computation?

File: trailmet/algorithms/prune/chipnet.py [hunk 5]
```diff
@@ -148,6 +155,7 @@ def compress_model(self):
             print('finetuning done')
```
Comment (dkgupta90) [inline trailmet/algorithms/prune/chipnet.py]: we need to do something about the print statements. May be use some sort of logger? If the print statement is not very important, remove it. We will later think of how to add the logger.

Comment (dkgupta90) [reply trailmet/algorithms/prune/chipnet.py]: generally, there is a parameter verbose=False or true that decides whether to print details or not. We can also use this parameter to control the level of information to be displayed.

File: trailmet/algorithms/prune/chipnet.py [hunk 6]
```diff
@@ -183,10 +191,7 @@ def prune(self, model, dataloaders, **kwargs):
                 remaining = self.get_remaining(self.steepness, self.budget_type).item()
                 remaining_before_pruning.append(remaining)
                 valid_accuracy.append(acc)
```
Comment (dkgupta90) [inline trailmet/algorithms/prune/chipnet.py:186]: ll things that are to be returned as output of the model compression process should be included in a a single dict.
logger['valid_accuracy'].append(acc)

File: trailmet/algorithms/prune/chipnet.py [hunk 7]
```diff
@@ -183,10 +191,7 @@ def prune(self, model, dataloaders, **kwargs):
...
-                # exactly_zeros, exactly_ones = model.plot_zt()
-                # exact_zeros.append(exactly_zeros)
-                # exact_ones.append(exactly_ones)
-                
+
                 print(f'[{epoch + 1} / {num_epochs}] Validation after pruning')
```
Comment (dkgupta90) [inline trailmet/algorithms/prune/chipnet.py:188]: if verbose:
    print(........)

File: trailmet/algorithms/prune/chipnet.py [hunk 8]
```diff
@@ -195,35 +200,37 @@ def prune(self, model, dataloaders, **kwargs):
...
                 # 
-                beta=min(6., beta+(0.1/self.b_inc))
-                gamma=min(256, gamma*(2**(1./self.g_inc)))
+                beta = min(6., beta + (0.1 / self.b_inc))
+                gamma = min(256, gamma * (2**(1. / self.g_inc)))
                 self.set_beta_gamma(beta, gamma)
-                print("Changed beta to", beta, "changed gamma to", gamma)     
-                
-                if acc>best_acc:
+                print("Changed beta to", beta, "changed gamma to", gamma)
```
Comment (dkgupta90) [inline trailmet/algorithms/prune/chipnet.py:201]: this should also be put under verbose.

File: trailmet/algorithms/prune/chipnet.py [hunk 9]
```diff
@@ -195,35 +200,37 @@ def prune(self, model, dataloaders, **kwargs):
...
                     }, f"checkpoints/{self.log_name}.pth")
 
                 df_data=np.array([remaining_before_pruning, remaining_after_pruning, valid_accuracy, pruning_accuracy, pruning_threshold, problems]).T
                 df = pd.DataFrame(df_data,columns = ['Remaining before pruning', 'Remaining after pruning', 'Valid accuracy', 'Pruning accuracy', 'Pruning threshold', 'problems'])
                 df.to_csv(f"logs/{self.log_name}.csv")
 
     def steepness_update_function(self, step):
+        """returns function to update steepness in budget loss of chipnet"""
         def update():
             self.steepness = min(60, self.steepness+step)
```
Comment (dkgupta90) [inline trailmet/algorithms/prune/chipnet.py]: add comment why the default is set to 60.

File: trailmet/algorithms/prune/chipnet.py [hunk 10]
```diff
@@ -254,27 +263,27 @@ def calculate_prune_threshold(self, target_budget, budget_type):
...
-                curr_budget+=zeta_weights[indx]
+            while curr_budget < (1. - target_budget):
+                indx += 1
+                curr_budget += zeta_weights[indx]
             prune_threshold = zetas[indx]
         else:
             prune_threshold = zetas[int((1.-target_budget)*len(zetas))]
         return prune_threshold
 
     def smoothRound(self, x, steepness=20.):
```
Comment (dkgupta90) [inline trailmet/algorithms/prune/chipnet.py:267]: docstring?

File: trailmet/algorithms/prune/chipnet.py [hunk 11]
```diff
@@ -254,27 +263,27 @@ def calculate_prune_threshold(self, target_budget, budget_type):
...
         return (module.pruned_zeta if module.is_pruned else self.smoothRound(module.get_zeta_t(), steepness)).sum()
-    
+
     def is_all_pruned(self, module):
         """checks if the whole block is pruned"""
         return self.n_remaining(module) == 0
-    
+
     def get_remaining(self, steepness=20., budget_type = 'channel_ratio'):
-        """return the fraction of active zeta_t (i.e > 0.5)"""
```
Comment (dkgupta90) [inline trailmet/algorithms/prune/chipnet.py:277]: more detailed docstring?

File: trailmet/algorithms/prune/chipnet.py [hunk 12]
```diff
@@ -400,12 +409,12 @@ def get_params_count(self):
...
+                total_params += total_param
             if isinstance(l_block, nn.Linear):
                 linear_params = l_block.weight.view(-1).shape[0]
-                active_params+=linear_params
-                total_params+=linear_params
+                active_params += linear_params
+                total_params += linear_params
         return active_params, total_params
 
     def get_volume(self):
```
Comment (dkgupta90) [inline trailmet/algorithms/prune/chipnet.py:414]: docstring?

**Review by dkgupta90 — CHANGES_REQUESTED** [review]

File: trailmet/algorithms/prune/chipnet.py (hunk 1 above)
Comment (rishabh-16) [reply trailmet/algorithms/prune/chipnet.py:30]: starting values for beta and gamma are always fixed and are not treated as hyperparameters, it is their increment values that matters and this is included in the chipnet_args.

File: trailmet/algorithms/prune/chipnet.py (hunk 4 above)
Comment (rishabh-16) [reply trailmet/algorithms/prune/chipnet.py:82]: not actually, its very specific to chipnet and prunablebatchnorm2d class

File: trailmet/algorithms/prune/chipnet.py (hunk 3 above)
Comment (rishabh-16) [reply trailmet/algorithms/prune/chipnet.py]: done

File: trailmet/algorithms/prune/chipnet.py (hunk 2 above)
Comment (rishabh-16) [reply trailmet/algorithms/prune/chipnet.py:55]: no it does not exist, but you are right about creating a projections class. what do you think where should we create it?

File: trailmet/algorithms/prune/chipnet.py (hunk 5 above)
Comment (rishabh-16) [reply trailmet/algorithms/prune/chipnet.py]: ok removing the print statements for now, will add logger later.

File: trailmet/algorithms/prune/chipnet.py (hunk 7 above)
Comment (rishabh-16) [reply trailmet/algorithms/prune/chipnet.py:188]: will add verbose and logger later

File: trailmet/algorithms/prune/chipnet.py (hunk 8 above)
Comment (rishabh-16) [reply trailmet/algorithms/prune/chipnet.py:201]: ok

File: trailmet/algorithms/prune/chipnet.py (hunk 9 above)
Comment (rishabh-16) [reply trailmet/algorithms/prune/chipnet.py]: done

File: trailmet/algorithms/prune/chipnet.py [hunk 13]
```diff
@@ -287,14 +296,14 @@ def get_remaining(self, steepness=20., budget_type = 'channel_ratio'):
...
                 k = l_block._conv_module.kernel_size[0]
                 prev_total = 3 if self.prev_module[l_block] is None else self.prev_module[l_block].num_gates
-                prev_remaining = 3 if self.prev_module[l_block] is None else self.n_remaining(self.prev_module[l_block], steepness) 
+                prev_remaining = 3 if self.prev_module[l_block] is None else self.n_remaining(self.prev_module[l_block], steepness)
                 n_rem += self.n_remaining(l_block, steepness)*prev_remaining*k*k
                 n_total += l_block.num_gates*prev_total*k*k
             elif budget_type == 'flops_ratio':
                 k = l_block._conv_module.kernel_size[0]
                 output_area = l_block._conv_module.output_area
                 prev_total = 3 if self.prev_module[l_block] is None else self.prev_module[l_block].num_gates
```
Comment (rishabh-16) [inline trailmet/algorithms/prune/chipnet.py:299]: reminder to resolve this error

File: trailmet/algorithms/prune/chipnet.py [hunk 14]
```diff
@@ -345,7 +354,7 @@ def prune_model(self, target_budget, budget_type = 'channel_ratio', finetuning=F
                 for l_block in self.prunable_modules:
                     l_block.prune(threshold)
                 self.remove_orphans()
-                if self.params()<target_budget:
+                if self.params() < target_budget:
```
Comment (rishabh-16) [inline trailmet/algorithms/prune/chipnet.py:351]: reminder to resolve this error

File: trailmet/algorithms/prune/chipnet.py [hunk 15]
```diff
@@ -359,15 +368,15 @@ def prune_model(self, target_budget, budget_type = 'channel_ratio', finetuning=F
                 for l_block in self.prunable_modules:
                     l_block.prune(threshold)
                 self.remove_orphans()
-                if self.flops()<target_budget:
+                if self.flops() < target_budget:
```
Comment (rishabh-16) [inline trailmet/algorithms/prune/chipnet.py:365]: reminder to resolve this error

File: trailmet/algorithms/prune/chipnet.py (hunk 13 above)
Comment (dkgupta90) [reply trailmet/algorithms/prune/chipnet.py:299]: what is this reminder?

**Review by dkgupta90 — APPROVED** [review]

**Review by Arnav0400 — APPROVED** [review]

**Pull request merged**